### PR TITLE
SJSU-1007 product search, SJSU-1010 Display individual category

### PR DESCRIPTION
**Pull request by minhphan156**
STORY: Create a search function
What I did:
-Add search bar and search button
-Created search page to show result
-Wired to allow navigations between pages
-Created back-end that allow to search part of string ( case in-sensitive)
-Created product model
To Verify:
_ Make sure you can navigate to other pages after the search page is showed
_ Uppercase and lowercase search key should give the same results

STORY: Display individual category
What I did:
_ New product card allow to show full name
_ all the cards is aligned neatly
_ Back-end for all 6 categories
_ Front-end is wired with Backend
_ Allow for Adding to card button
_ fixed bug with Vinny code
To Verify:
_ test all categories by clicking

### CODE REVIEW

File: front_end/src/App.css [hunk 1]
```diff
@@ -2,13 +2,20 @@ img {
...
+}
+
 .landing {
   position: relative;
   background-size: cover;
   background-position: center;
   height: 100vh;
   margin-top: -24px;
   margin-bottom: -50px;
+  /* The image used */
```
Comment (SandroSal) [inline front_end/src/App.css]: it's not really an image... it's a gif :)

File: front_end/src/App.js [hunk 2]
```diff
@@ -15,14 +15,15 @@ import Footer from "./components/layout/Footer";
 import Landing from "./components/layout/Landing";
 import Register from "./components/auth/Register";
 import Login from "./components/auth/Login";
+import Search from "./components/layout/Search";
 import Dashboard from "./components/dashboard/Dashboard";
 import CreateProfile from "./components/create-profile/CreateProfile";
 import EditProfile from "./components/edit-profile/EditProfile";
-import ProductDetail from "./components/product/ProductDetail";
 import BrowseByCategory from "./components/browse/BrowseByCategory";
+import categoryPage from "./components/browse/categoryPage";
```
Comment (SandroSal) [inline front_end/src/App.js]: please make this capital

Comment (SandroSal) [reply front_end/src/App.js]: CategoryPage

**Comment by DaleCS** [comment]
Code looks good, works well, no merge conflicts, branched off dev, and follows both stories closely.

File: front_end/src/actions/categoryActions.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import { GET_CATEGORY } from "./types";
+import axios from "axios";
+// 2-do post request at routes/product.js with newQuery object
```
Comment (SandroSal) [inline front_end/src/actions/categoryActions.js]: what are those numbers for?

File: front_end/src/actions/categoryActions.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+export const submitCategory = newCategory => dispatch => {
+  axios
+    .get("/api/product/category/", { params: newCategory })
+    ///api/product/search/?description=searchString
```
Comment (SandroSal) [inline front_end/src/actions/categoryActions.js]: comments necessary?

File: front_end/src/actions/queryActions.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import { SET_QUERY } from "./types";
+import axios from "axios";
+// 2-do post request at routes/product.js with newQuery object
```
Comment (SandroSal) [inline front_end/src/actions/queryActions.js]: needed?

File: front_end/src/actions/queryActions.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+// newQuery = {description : someString }
+
+export const submitQuery = newQuery => dispatch => {
+  axios
+    .get("/api/product/search/", { params: newQuery })
+    ///api/product/search/?description=searchString
+    // 4-if success then res.data will be the product object we looking for
```
Comment (SandroSal) [inline front_end/src/actions/queryActions.js]: needed?

File: front_end/src/components/browse/BrowseByCategory.js [hunk 7]
```diff
@@ -1,7 +1,23 @@
...
 
 class BrowseByCategory extends Component {
+  constructor() {
+    super();
+    this.onClick = this.onClick.bind(this);
+  }
+
+  onClick(e) {
+    //NOTE: we assume user will search for name
+    // 1-submit query as object with to submitQuery at queryActions.js
```
Comment (SandroSal) [inline front_end/src/components/browse/BrowseByCategory.js]: what's the 1 for?

File: front_end/src/components/browse/BrowseByCategory.js [hunk 8]
```diff
@@ -11,54 +27,83 @@ class BrowseByCategory extends Component {
         </div>
         <div className="row justify-content-md-center">
           <div className="col-sm-2 border category-tile shadow">
-            <Link to="/categories">
-              <img src="category-imgs/Snacks_c.png" className="category-img" alt="Snacks.img"/>
+            <Link to="/Snack" onClick={() => this.onClick("snack")}>
+              <img
+                src="category-imgs/Snacks_c.png"
+                className="category-img"
```
Comment (SandroSal) [inline front_end/src/components/browse/BrowseByCategory.js]: just for my understanding..
                className="category-img"
is used to make all the pics have the same size etc?

File: front_end/src/components/browse/BrowseByCategory.js [hunk 9]
```diff
@@ -67,4 +112,11 @@ class BrowseByCategory extends Component {
   }
 }
 
-export default BrowseByCategory;
+const mapStateToProps = state => ({});
+// this.props.query = { productQuery : data}
```
Comment (SandroSal) [inline front_end/src/components/browse/BrowseByCategory.js]: needed?

File: front_end/src/components/browse/categoryPage.js [hunk 10]
```diff
@@ -0,0 +1,44 @@
+import React, { Component } from "react";
```
Comment (SandroSal) [inline front_end/src/components/browse/categoryPage.js]: Please either rename the entire file to something more descriptive as "CategoryShowProducts"

or add a comment at the beginning of the file, explaining what it is

File: front_end/src/components/browse/BrowseByCategory.js (hunk 9 above)
Comment (DaleCS) [inline front_end/src/components/browse/BrowseByCategory.js]: Is this a to-do comment for future functionality?

File: front_end/src/components/browse/categoryPage.js [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+    return <div />;
+  }
+}
+
+const mapStateToProps = state => ({
+  category: state.category
+});
+// this.props.category = { productcategory : data}
+// if this.props.category is empty we will not show the categoryPage page
```
Comment (SandroSal) [inline front_end/src/components/browse/categoryPage.js]: what are those comments for?

File: front_end/src/components/browse/categoryPage.js [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    return <div />;
+  }
+}
+
+const mapStateToProps = state => ({
+  category: state.category
+});
+// this.props.category = { productcategory : data}
```
Comment (DaleCS) [inline front_end/src/components/browse/categoryPage.js]: Another one

File: front_end/src/components/layout/Search.js [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    return <div />;
+  }
+}
+
+const mapStateToProps = state => ({
+  query: state.query
+});
+// this.props.query = { productQuery : data}
```
Comment (DaleCS) [inline front_end/src/components/layout/Search.js]: Another one.

File: front_end/src/components/browse/categoryPage.js [hunk 14]
```diff
@@ -0,0 +1,44 @@
+import React, { Component } from "react";
+import { PropTypes } from "prop-types";
+import { connect } from "react-redux";
+import ProductCard from "../product-tiles/ProductCard";
+
+class categoryPage extends Component {
+  render() {
+    const category = this.props.category.productCategory;
```
Comment (SandroSal) [inline front_end/src/components/browse/categoryPage.js]: just for my understanding, what is "productCategory"? and where is it defined? It must be an array of products, right?

Comment (SandroSal) [reply front_end/src/components/browse/categoryPage.js]: nvm, found it in the reducer

File: front_end/src/reducers/index.js [hunk 15]
```diff
@@ -1,13 +1,19 @@
 import { combineReducers } from "redux";
 import authReducer from "./authReducer";
 import errorReducer from "./errorReducer";
+import queryReducer from "./queryReducer";
 import profileReducer from "./profileReducer";
+import categoryReducer from "./categoryReducer";
 import productReducer from "./productReducer";
 
 // authReducer = { new initialState }
```
Comment (DaleCS) [inline front_end/src/reducers/index.js]: Commented out unused variables?

File: routes/api/product.js [hunk 16]
```diff
@@ -1,12 +1,60 @@
...
-const mongoose = require("mongoose");
-const passport = require("passport");
 
 // Load Product model
 const Product = require("../../models/Product");
 
+// @route GET api/product/search
+// @desc Register user
+// @access Public
+// router.get("/search/:description", (req, res) => {
```
Comment (DaleCS) [inline routes/api/product.js]: Another to-do comment?

**Comment by DaleCS** [comment]
Another than the questions that I have, the code works well, no merge conflicts, it's branched off dev, and follows the stories closely.

File: front_end/src/components/layout/Navbar.js [hunk 17]
```diff
@@ -17,24 +44,26 @@ class Navbar extends Component {
...
           </Link>
         </li>
-
         <li className="nav-item">
           <a
             href="#"
             onClick={this.onLogoutClick.bind(this)}
             className="nav-link"
           >
+            <img
```
Comment (SandroSal) [inline front_end/src/components/layout/Navbar.js]: this should not be part of logout. the image is misleading

File: front_end/src/components/layout/Navbar.js [hunk 18]
```diff
@@ -43,6 +72,9 @@ class Navbar extends Component {
 
     const guestLinks = (
       <ul className="navbar-nav ml-auto">
+        <Link to="/cart" className="btn btn-light">
```
Comment (SandroSal) [inline front_end/src/components/layout/Navbar.js]: is there a reason to have the cart for authenticated users and for guests separately?

File: front_end/src/components/layout/Navbar.js [hunk 19]
```diff
@@ -81,10 +136,13 @@ Navbar.propTypes = {
 };
 
 const mapStateToProps = state => ({
-  auth: state.auth
+  auth: state.auth,
+  query: state.query
 });
+// this.props.query = { productQuery : data}
+// if this.props.query is empty we will not show the Search page
```
Comment (SandroSal) [inline front_end/src/components/layout/Navbar.js]: needed?

File: front_end/src/components/layout/Search.js [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+// if this.props.query is empty we will not show the Search page
```
Comment (SandroSal) [inline front_end/src/components/layout/Search.js]: needed?

File: front_end/src/components/product-tiles/ProductCard.js [hunk 21]
```diff
@@ -0,0 +1,36 @@
+import React, { Component } from "react";
```
Comment (SandroSal) [inline front_end/src/components/product-tiles/ProductCard.js]: The same file (with different content) is in Dale's PR

File: front_end/src/reducers/index.js [hunk 22]
```diff
@@ -1,13 +1,19 @@
...
+// query = { productQuery : data}
```
Comment (SandroSal) [inline front_end/src/reducers/index.js]: what is this for?

File: models/Product.js [hunk 23]
```diff
@@ -38,4 +38,4 @@ const productSchema = new Schema({
   }
 });
 
-module.exports = Product = mongoose.model("Product", productSchema); // create model database
\ No newline at end of file
+module.exports = Product = mongoose.model("Product", productSchema); // create model database
```
Comment (SandroSal) [inline models/Product.js]: why is this a change?

File: routes/api/product.js [hunk 24]
```diff
@@ -1,12 +1,60 @@
...
+router.get("/search/", (req, res) => {
+  // 3-receive req to query from submitQuery at /actions/queryActions.js
```
Comment (SandroSal) [inline routes/api/product.js]: I don't really get this part

File: routes/api/product.js [hunk 25]
```diff
@@ -1,12 +1,60 @@
...
+    .catch(err => {
+      console.error(err);
+    });
+});
+
+// @route GET api/product/category
+// @desc Register user
+// @access Public
+router.get("/category/", (req, res) => {
+  // 3-receive req to query from submitcategory at /actions/categoryActions.js
```
Comment (SandroSal) [inline routes/api/product.js]: again, what does that mean

File: front_end/src/components/product-tiles/ProductCard.js (hunk 21 above)
Comment (DaleCS) [reply front_end/src/components/product-tiles/ProductCard.js]: I assume his productCard is a placeholder for code testing purposes.

Comment (minhphan156) [reply front_end/src/components/product-tiles/ProductCard.js]: this is the original product card used at DEMO . Dale's new PR is the improved one. I think we should merge this first

File: models/Product.js (hunk 23 above)
Comment (minhphan156) [reply models/Product.js]: i just copy and paste this whole file from DEV . I am trying not to miss anything

File: front_end/src/components/browse/BrowseByCategory.js (hunk 8 above)
Comment (DaleCS) [reply front_end/src/components/browse/BrowseByCategory.js]: Yes. It's going to be changed in my story.

File: front_end/src/components/browse/categoryPage.js (hunk 10 above)
Comment (minhphan156) [reply front_end/src/components/browse/categoryPage.js]: renamed to CategoryShowProducts

File: front_end/src/components/layout/Navbar.js (hunk 17 above)
Comment (minhphan156) [reply front_end/src/components/layout/Navbar.js]: removed img tag

File: front_end/src/components/layout/Navbar.js (hunk 19 above)
Comment (minhphan156) [reply front_end/src/components/layout/Navbar.js]: removed the top one

File: front_end/src/components/layout/Search.js (hunk 20 above)
Comment (minhphan156) [reply front_end/src/components/layout/Search.js]: removed the top one

File: front_end/src/components/layout/Navbar.js (hunk 18 above)
Comment (minhphan156) [reply front_end/src/components/layout/Navbar.js]: moved out of the users

**Pull request merged**